docs/compilers/CSharp/Nullability Public API Design Notes.md

```diff
@@ -0,0 +1,182 @@
+# Nullability API Design Notes
+
+## Guiding Principles
+
+When designing this API, there were a few core principles that we tried to keep in mind.
+
+1. First and foremost, we do not want to break existing analyzers and refactorings. This means that common patterns that used to work need to continue to work. These include reference equality for `ITypeSymbol`s that aren't generics, such as `String`.
+2. We'd like the ability to avoid computing nullability information if it's not required. Currently, the SemanticModel APIs do a good job of only causing the minimal amount of binding required to answer the question being asked, but attempting to compute nullability information will cause significantly more to need to be bound. To fully determine nullability, we must bind the entire method and run nullability analysis up to the point requested by the caller, unlike today where individual statements can get away with being bound by themselves. Even if the standard path calculates nullability info, we feel that some scenarios will be both perf sensitive and not care about results of nullability analysis. This includes things like intellisense.
+3. Users should be able to get at both declared nullability for variables, as well as seeing the current flow state of expressions based on flow analysis.
+
+## General Concepts
+
+Nullability information is exposed publicly as an enum, `Nullability`. There are 5 possible values:
+
+* `NotApplicable` - This is used when nullability information is requested about something that does not have a nullability, such as a statement, directive, or other similar constructs.
+* `NotAnalyzed` - This is used to fulfill priority #2. If the user requests information while asking for nullability to not be calculated, we'll use this to stand in for nullability.
+* `Unknown` - This is used for legacy C# code that does not know about nullability. This can appear in modern C# code that calls unannoted C#. For use in this document, we represent unknown via `~`, but this symbol does not appear in source.
+* `NotNull` - This is used when we know the nullability is not null.
+* `MaybeNull` - This is used when we know the nullability might be null.
+
+Nullability information will be retreived from the `SemanticModel`, much like type information. `GetSemanticModel` will return a `SemanticModel` that is aware of nullability information, and will force binding and calculation of nullability information when queried. We will offer a method off of `SemanticModel`, `GetSingleStatementSemanticModel()`, which will return a `SemanticModel` that does not calculate nullability information, for particularly performance-sensitive scenarios. This `SemanticModel` will share a cache with its parent, much like speculative `SemanticModel`s, and all nullabilities accessed on symbols retrieved from this `SemanticModel` will be `NotAnalyzed`.
```

Comment:
How are we documenting GetSingleStatementSemanticModel? Is it strictly stated "it just doesn't compute nullability" or are we leaving the door open for other flow things like this down the road?

Reply:
No longer relevant.

---
In reply to: [240430403](https://github.com/333fred/roslyn/pull/1#discussion_r240430403) [](ancestors = 240430403)